staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -974,28 +986,45 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
 	expectedRV := 2000
 
 	wg := sync.WaitGroup{}
+	done := make(chan struct{})
+	errorCh := make(chan error, 1)
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
 		for {
-			event, ok := <-w.ResultChan()
-			if !ok {
-				t.Fatalf("Unexpected closed channel")
-			}
-			rv, err := cacher.versioner.ObjectResourceVersion(event.Object)
-			if err != nil {
-				t.Errorf("failed to parse resource version from %#v: %v", event.Object, err)
-			}
-			if event.Type == watch.Bookmark && rv == uint64(expectedRV) {
+			select {
+			case <-done:
 				return
+			default:
+				event, ok := <-w.ResultChan()
+				if !ok {
+					errorCh <- fmt.Errorf("Unexpected closed channel")
+					return
+				}
+				rv, err := cacher.versioner.ObjectResourceVersion(event.Object)
+				if err != nil {
+					errorCh <- fmt.Errorf("failed to parse resource version from %#v: %v", event.Object, err)
+					return
+				}
+				if event.Type == watch.Bookmark && rv == uint64(expectedRV) {
+					return
+				}
 			}
 		}
 	}()
 
 	// Simulate progress notify event.
 	cacher.watchCache.UpdateResourceVersion(strconv.Itoa(expectedRV))
 
-	wg.Wait()
+	go func() {
```

Comment:
Same here, not sure why a new go routine is needed.

Reply:
Same as above, to avoid blocking and allow receive on `if err = <- errorCh; err != nil {...}`